pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go

```diff
@@ -2886,3 +3152,38 @@ func GetLoadBalancerSourceRanges(service *corev1.Service) (netsets.IPNet, error)
 	}
 	return ipnets, nil
 }
+
+// PreserveGopherError preserves the error details delivered with the response
+// that are explicitly omitted by dedicated error types.
+// error types from provider_client.go
+func PreserveGopherError(rawError error) error {
+	if rawError == nil {
+		return nil
+	}
+	var details []byte
+	switch e := rawError.(type) {
+	case gophercloud.ErrDefault400:
+	case gophercloud.ErrDefault401:
+		details = e.Body
+	case gophercloud.ErrDefault403:
+	case gophercloud.ErrDefault404:
+		details = e.Body
+	case gophercloud.ErrDefault405:
+		details = e.Body
+	case gophercloud.ErrDefault408:
```

Comment:
Don't you want to add a support for `gophercloud.ErrDefault409`? FIP allocation may return two types of code with a body:

400:
```
{
  "NeutronError": {
    "detail": "",
    "message": "Unable to find any IP address on external network 99e8a528-0b85-419a-9f7f-ee59a92dc7a2.",
    "type": "ExternalIpAddressExhausted"
  }
}
```

409:
```
{
  "NeutronError": {
    "detail": "",
    "message": "No more IP addresses available on network 99e8a528-0b85-419a-9f7f-ee59a92dc7a2.",
    "type": "IpAddressGenerationFailure"
  }
}
```

Reply:
See how it's done in https://github.com/terraform-provider-openstack/terraform-provider-openstack/pull/1163/files
and https://github.com/terraform-provider-openstack/terraform-provider-openstack/pull/1171/files

Reply:
Error 409 implementation already does not hide the error details. I've added only those errors here, where its implementation overrides the informative base Error implementation with a simple text hiding the details (what even the reason for this might be).

Should I add all errors here, regardless of its standard implementation?

Reply:
Yes, I added the others also, with a default behaviour, therefore I added now the missed 409